Develop

yixwellet opened the pull request.

yixwellet committed:
установлены зависимости, настроены сборки

yixwellet committed:
настроено подключение к монго, созданы схемы и модели

yixwellet committed:
настроено подключение к монго, созданы схемы и модели

yixwellet committed:
созданы контроллеры и роуты для добавления/просмотра юзеров, для добавления/удаления/просмотра карточек

yixwellet committed:
реализовано обновление профиля и аватара, лайки и дизлайки карточкам

yixwellet committed:
мелкие доработки

talantbekov123 commented:
Хорошая работа, пользователи и картчоки создаются, есть возможность получить пользователя по его id, подключение к базе происходит по корректному адресу, так же хочу отметить хорошие знания Es6 (стрелочные функции) работа с базой через промисы а не коллбэки! У проекта ясная структура, код разбит на роуты, модели и контроллеры.
Прошу исправить комментарии надо исправить что бы я смог принять работу.


**Надо исправить**
- Ошибки, связанные с невалидными данными, переданными на сервер, лучше сопровождать статусом 400.Например если некорректная ссылка или некорректное имя пользователя. Вместо этого приходит ответ 500 с общей ошибкой.  Подробнее [тут](https://developer.mozilla.org/ru/docs/Web/HTTP/Status/400)
- При попытке удаления карточки, ID которой отсутствует в БД сервис возвращает
```
{
   "data": null
}
```
- API возвращает ответ `null ` со статусом 200 при попытке получить юзера по валидному ObjectId "5f213f2beb515119340e1111", которого не существует в базе. Если пользователя или карточки с запрошенным идентификатором нет, API должен возвращать обьект с ключем message и статус 404;


**Можно лучше**
- Валидацию можно делать при помощи функции const validator = require('validator'); Пример ниже
```
link: {
    type: String,
    validate: {
      validator(link) {
        return validator.isURL(link);
      },
    },
    required: true,
  },
```

yixwellet committed:
исправлены замечания

talantbekov123 commented:
Хорошая работа, комментарии исправлены. Вдвойне похвально что выполнили комментариям можно лучше. Есть комментарии которые нужно исправить что бы я мог принять работу.

**Надо исправить**
- Лайк/дизлайк карточки которой нет в базе вернул ответ null со статусом 200, прошу проверить. Пример запроса PUT `/cards/5f04c2fe6dcb981178b2b75b/likes`. Должны получить json c полем message с статусом 404 (Bad Request).
- Запросы с невалидными данными должны сопровождаться статусом 400. На данный момент ответы на невалидные данные сопровождаются статусом 500. Пример не коректного статуса по ссылке(https://ibb.co/s9rrTQ2.). При создании пользователя(POST /users) прошу исправить.
- Нет проверки на наличие пользователя в базе перед обновлением. Если в app.js задать url которого нет в базе, ответ успешный. `req.user = { _id: '5f2bf00b1125ef2608a11111', };`

**Можно лучше**
- С комитами все отлично, но можно немного лучше. Часто возникает соблазн написать комментарий к коммиту вида "fixes" или "first commit" или "исправления". Лучше описывать коммиты подробнее.
Делать коммиты, когда:
1) выполнил единицу работы
2) сделал изменения, которые потом может потребоваться отменить
Интересная статья по [ссылке](https://habr.com/ru/company/Voximplant/blog/276695/)

yixwellet committed:
Исправлены замечания:
1. Выводится ошибка при попытке обращения к несуществующей карточке
2. Выводится ошибка при попытке обновить данные несуществующего пользователя
3. На запросы с данными, не прошедшими валидацию, возвращается ответ со статусом 400

talantbekov123 commented:
Все проверил повторно, отличная работа. Единственное заметил что, остался 1 не исправленный комментарий с 1 вой итерации. Прошу исправить, постараюсь сделать проверку максимально быстро.

**Надо исправить**
- При попытке удаления карточки, ID которой отсутствует в БД сервис возвращает
```
{
   "data": null
}
```

yixwellet committed:
При попытке удаления несуществующей карточки возвращается ошибка 404

yixellet commented:
Спасибо за подсказку насчет orFail(), но к сожалению я не могу заставить его работать. Сдаю работу без него, чтобы не тянуть время, но orFail() изучать продолжаю.

talantbekov123 commented:
Отличная работа, все комментарии исправлены! Похвально! Принимаю.
Чуть позже вышлю вам пример работы с orFail()

talantbekov123 commented:
Отличная работа, все комментарии исправлены! Похвально! Принимаю.
